torch/nn/utils/parametrize.py

```diff
@@ -53,63 +55,137 @@ def cached():
-    a parametrized :class:`torch.nn.Module`. It is the type of
-    ``module.parametrizations[tensor_name]`` when ``module[tensor_name]`` has been parametrized
-    with :func:`register_parametrization`.
+    r"""A sequential container that holds and manages the ``original`` or ``original0``, ``original1``, ...
+    parameters or buffers of a parametrized :class:`torch.nn.Module`.
+
+    It is the type of ``module.parametrizations[tensor_name]`` when ``module[tensor_name]``
+    has been parametrized with :func:`register_parametrization`.
+
+    If the first registered parmetrization has a ``right_inverse`` that returns one tensor or
+    does not have a ``right_inverse`` (in which case we assume that ``right_inverse`` is the identity),
+    it will hold the tensor under the name ``original``.
+    If it has a ``right_inverse`` that returns more than one tensor, these will be registered as
+    ``original0``, ``original1``, ...
 
     .. note ::
         This class is used internally by :func:`register_parametrization`. It is documented
         here for completeness. It should not be instantiated by the user.
 
     Args:
-        modules (iterable): an iterable of modules representing the parametrizations
+        modules (sequence): an sequence of modules representing the parametrizations
         original (Parameter or Tensor): parameter or buffer that is parametrized
     """
     original: Tensor
 
     def __init__(
-        self, modules: Iterable[Module], original: Union[Tensor, Parameter]
+        self, modules: Sequence[Module], original: Union[Tensor, Parameter]
     ) -> None:
-        super().__init__(modules)
-        if isinstance(original, Parameter):
-            self.register_parameter("original", original)
-        else:
-            self.register_buffer("original", original)
+        # We require this because we need to treat differently the first parametrization
+        # This should never throw, unless this class is used from the outside
+        if len(modules) == 0:
+            raise ValueError("ParametrizationList requires one or more modules.")
 
-    def set_original_(self, value: Tensor) -> None:
-        r"""This method is called when assigning to a parametrized tensor.
+        super().__init__(modules)
 
-        It calls the methods ``right_inverse`` (see :func:`register_parametrization`)
+        with torch.no_grad():
+            value = original
+            for module in reversed(self):  # type: ignore[call-overload]
+                if hasattr(module, "right_inverse"):
+                    value = module.right_inverse(value)
+                # else, we assume that it's the identity
+
+        if not isinstance(value, Tensor) and not isinstance(value, collections.Sequence):
+            raise ValueError("'right_inverse' should return a Tensor or a Sequence of tensors (list, tuple...). "
+                             f"Got {value.__class__.__name__}")
+
+        # Set the number of original tensors
```

Comment:
A common trick we use here is to get value to always be a sequence by wrapping single Tensors into a tuple of size 1.
- That simplifies the logic here as you don't need to special case for single Tensor
- You can always handle `original0` instead of `original` when there is a single Tensor
- Calling into the forward can be done with common code as well.
- If you really need to unpack it into a single Tensor (for the use of set_ for example), you can always save a boolean along with the sequence

Reply:
I tried doing that, but the treatment of these two cases is actually different enough to not be worth it---it clutters too much the implementation.

This comes from the fact that we reuse the tensor to avoid changing the `id` of the parameter and so on. Note, for example, that the exceptions raised in the if-else below are fundamentally different.